File: src/libpumpkin/Ctl.zig
const pumpkin = @import("pumpkin.zig");
const ControlType = pumpkin.ControlType;

const c = @cImport({
  @cInclude("zigpumpkin.h");
});

pub fn showControl(controlP: *ControlType) void {
  c.CtlShowControl(controlP);
}

pub fn hideControl(controlP: *ControlType) void {
  c.CtlHideControl(controlP);
}

pub fn getValue(controlP: *ControlType) u16 {
  return c.CtlGetValue(controlP);
}

pub fn setValue(controlP: *ControlType, newValue: u16) void {
  c.CtlSetValue(controlP, newValue);
}

pub fn setSliderValues(controlP: *ControlType, minValue: *u16, maxValue: *u16, pageSize: *u16, value: *u16) void {
  c.CtlSetSliderValues(controlP, minValue, maxValue, pageSize, value);
}

pub fn getSliderValues(controlP: *ControlType, minValue: *u16, maxValue: *u16, pageSize: *u16, value: *u16) void {
  c.CtlGetSliderValues(controlP, minValue, maxValue, pageSize, value);
}

pub fn getLabel(controlP: *ControlType) [*]const u8 {
  return c.CtlGetLabel(controlP);
}

pub fn setLabel(controlP: *ControlType, newLabel: [*]const u8) void {
  c.CtlSetLabel(controlP, newLabel);
}

pub fn updateGroup(controlP: *ControlType, value: bool) void {
  c.CtlUpdateGroup(controlP, if (value) 1 else 0);
}

pub fn setGraphics(controlP: *ControlType, newBitmapID: u16, newSelectedBitmapID: u16) void {
  c.CtlSetGraphics(controlP, newBitmapID, newSelectedBitmapID);
}

File: src/libpumpkin/ctl.c
inline void CtlDrawControl(ControlType *ctl) {
  pumpkin_system_call(sysTrapCtlDrawControl, 0, NULL, NULL, ctl);
}

inline void CtlEraseControl(ControlType *ctl) {
  pumpkin_system_call(sysTrapCtlEraseControl, 0, NULL, NULL, ctl);
}

inline void CtlHideControl(ControlType *ctl) {
  pumpkin_system_call(sysTrapCtlHideControl, 0, NULL, NULL, ctl);
}

inline void CtlShowControl(ControlType *ctl) {
  pumpkin_system_call(sysTrapCtlShowControl, 0, NULL, NULL, ctl);
}

inline Int16 CtlGetValue(const ControlType *ctl) {
  uint64_t iret;
  pumpkin_system_call(sysTrapCtlGetValue, 0, &iret, NULL, ctl);
  return (Int16)iret;
}

inline void CtlSetValue(ControlType *ctl, Int16 value) {
  pumpkin_system_call(sysTrapCtlSetValue, 0, NULL, NULL, ctl, value);
}

inline const char *CtlGetLabel(const ControlType *ctl) {
  void *pret;
  pumpkin_system_call(sysTrapCtlGetLabel, 0, NULL, &pret, ctl);
  return (char *)pret;
}

inline void CtlSetLabel(ControlType *ctl, const char *text) {
  pumpkin_system_call(sysTrapCtlSetLabel, 0, NULL, NULL, ctl, text);
}

inline Boolean CtlHandleEvent(ControlType *ctl, EventType *event) {
  uint64_t iret;
  pumpkin_system_call(sysTrapCtlHandleEvent, 0, &iret, NULL, ctl, event);
  return (Boolean)iret;
}

inline void CtlHitControl(const ControlType *ctl) {
  pumpkin_system_call(sysTrapCtlHitControl, 0, NULL, NULL, ctl);
}

inline void CtlSetEnabled(ControlType *ctl, Boolean enabled) {
  pumpkin_system_call(sysTrapCtlSetEnabled, 0, NULL, NULL, ctl, enabled);
}

inline void CtlSetUsable(ControlType *ctl, Boolean usable) {
  pumpkin_system_call(sysTrapCtlSetUsable, 0, NULL, NULL, ctl, usable);
}

inline Boolean CtlEnabled(const ControlType *ctl) {
  uint64_t iret;
  pumpkin_system_call(sysTrapCtlEnabled, 0, &iret, NULL, ctl);
  return (Boolean)iret;
}

inline void CtlSetGraphics(ControlType *ctl, DmResID newBitmapID, DmResID newSelectedBitmapID) {
  pumpkin_system_call(sysTrapCtlSetGraphics, 0, NULL, NULL, ctl, newBitmapID, newSelectedBitmapID);
}

inline void CtlGetSliderValues(const ControlType *ctl, UInt16 *minValue, UInt16 *maxValue, UInt16 *pageSize, UInt16 *value) {
  pumpkin_system_call(sysTrapCtlGetSliderValues, 0, NULL, NULL, ctl, minValue, maxValue, pageSize, value);
} 
  
inline void CtlSetSliderValues(ControlType *ctl, const UInt16 *minValue, const UInt16 *maxValue, const UInt16 *pageSize, const UInt16 *value) {
  pumpkin_system_call(sysTrapCtlSetSliderValues, 0, NULL, NULL, ctl, minValue, maxValue, pageSize, value);
}

inline Boolean CtlValidatePointer(const ControlType *ctl) {
  uint64_t iret;
  pumpkin_system_call(sysTrapCtlValidatePointer, 0, &iret, NULL, ctl);
  return (Boolean)iret;
} 
